src/methods/join.js

```diff
@@ -28,7 +28,37 @@ import { AbstractValue, Value } from "../values/index.js";
 import invariant from "../invariant.js";
 import * as t from "babel-types";
 
-export function joinPossiblyNormalCompletions(
+export function unbundleNormalCompletion(
+  completionOrValue: Completion | Value | Reference
+): [void | NormalCompletion, Value | Reference] {
+  let completion, value;
+  if (completionOrValue instanceof NormalCompletion) {
+    completion = completionOrValue;
+    value = completionOrValue.value;
+  }
+  return [completion, value];
+}
+
+export function composeNormalCompletions(
+  leftCompletion: void | Completion | Value, rightCompletion: void | Completion | Value, resultValue: Value, realm: Realm
+): PossiblyNormalCompletion | Value {
+  if (leftCompletion instanceof PossiblyNormalCompletion) {
+    if (rightCompletion instanceof PossiblyNormalCompletion) {
+      updatePossiblyNormalCompletionWithValue(realm, rightCompletion, resultValue);
+      return composePossiblyNormalCompletions(realm, leftCompletion, rightCompletion);
+    }
+    updatePossiblyNormalCompletionWithValue(realm, leftCompletion, resultValue);
+    return leftCompletion;
+  } else if (rightCompletion instanceof PossiblyNormalCompletion) {
+    updatePossiblyNormalCompletionWithValue(realm, rightCompletion, resultValue);
+    return rightCompletion;
+  } else {
+    invariant(!leftCompletion && !rightCompletion);
```

Comment:
It's not really clear what the requirements on the arguments are from this invariant. It seems that both left and right should be NormalCompletions, but that doesn't seem to be enforced in any way.